### PR TITLE
Optimized getMetricsWithLabelNames() helper

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does**:
While profiling the store-gateway CPU and memory in a cluster running with a large number of tenants per store-gateway instance, I've noticed a small (but significative) % of CPU and allocated objects are due to `getMetricsWithLabelNames()`. In this PR I'm proposing some optimization.

Benchmark:

```
name                         old time/op    new time/op    delta
GetMetricsWithLabelNames-12     931µs ± 1%     225µs ± 2%  -75.86%  (p=0.008 n=5+5)

name                         old alloc/op   new alloc/op   delta
GetMetricsWithLabelNames-12     735kB ± 0%      89kB ± 0%  -87.88%  (p=0.008 n=5+5)

name                         old allocs/op  new allocs/op  delta
GetMetricsWithLabelNames-12     4.03k ± 0%     2.01k ± 0%  -50.10%  (p=0.008 n=5+5)
```

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/metrics_helper_test.go [hunk 1]
```diff
@@ -95,6 +95,45 @@ func TestGetMetricsWithLabelNames(t *testing.T) {
 	}, out2)
 }
 
+func BenchmarkGetMetricsWithLabelNames(b *testing.B) {
+	const (
+		numMetrics         = 1000
+		numLabelsPerMetric = 10
+	)
```
Comment (jtlisi) [inline pkg/util/metrics_helper_test.go:102]: **thought:** I'd be interested to see how the benchmark results change if you increase these values a bit. I'd imagine the effect is amplified for higher cardinality queries

**Review by jtlisi — APPROVED** [review]
LGTM, good find

**Review by gotjosh — APPROVED** [review]

File: pkg/util/metrics_helper_test.go (hunk 1 above)
Comment (pracucci) [reply pkg/util/metrics_helper_test.go:102]: > higher cardinality queries

This helper is not used by queries but to aggregate metrics exposed by ` /metrics`. 1K for a single aggregation is below the number in the cluster I'm profiling, but above the average in most Cortex installations.

**Review by pstibrany — APPROVED** [review]
Nice improvement!

**Pull request merged**